picard/album.py

```diff
@@ -353,13 +351,11 @@ def _recordings_request_finished(self, document, http, error):
             offset = document.get('recording-offset', 0)
             next_offset = offset + RECORDING_QUERY_LIMIT
             if next_offset < count:
-                self._requests -= 1
                 self._request_recording_relationships(offset=next_offset)
             else:
                 # Merge separately loaded recording relationships into release node
                 self._merge_release_recording_relationships()
                 self._run_album_metadata_processors()
-                self._requests -= 1
                 self._finalize_loading(error)
```

Comment:
Damn, good catch, we really need to find a better way, managing was +/-1 is very prone to errors.

One approach would be to use decorators and internal counters:

```python
@increment_here
def A(....):
    B(...)

@decrement_here
def B(...):
   ....

def C(...):
  try:
     A()
   except SomethingWentWrongWithCounters:
     handle_the_case()
 
   

````


Anyway, commit removed, PR rebased

Reply:
[PICARD-1403](https://tickets.metabrainz.org/browse/PICARD-1403) :)

I think just with decorators won't do. This must be usable by plugins as well. What I think should happen is that we can register requests on an album. Something like "album.register_request(request)".

One problem is when to decrement the counter again. It could be done automatically when the request finishes. But usually you want it to happen **after** the handler registered by the plugin has run. Otherwise you become side effects when `_finalize_loading` gets called and the counter is already 0, but one plugin is still busy doing stuff.

One advantage of registering requests with the album would be that it allows implementing PICARD-1507.